### PR TITLE
Fix column reordering in process_text_na_dataframe  with nan and missing values due to Columntransformer

### PR DESCRIPTION
**Pull request by MagnusBuehler**
## Motivation and Context

- fixing bug mentioned in _https://github.com/PriorLabs/TabPFN/issues/530_
- columntransformer does not keep the feature order
- therefore, this change computes a mapping from the new ordering (not what we keep) to the original column order
- added tests to check different use-cases with nan and missing values

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

Locally. Implemented a test, which runs different combinations of nan and missing values. 

## Checklist

-   [x] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to `CHANGELOG.md` (if relevant for users).
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/tabpfn/utils.py [hunk 1]
```diff
@@ -627,6 +627,22 @@ def process_text_na_dataframe(
     else:
         X_encoded = X
 
+    if ord_encoder is not None:
+        # The ColumnTransformer will reorder the columns as following:
+        # - categorical columns (in the order they appear in the dataframe)
+        # - numerical columns (in the order they appear in the dataframe)
+        # We need to revert this to the original order.
+        # assumes the column transformer has one OrdinalEncoder, which is
+        # what we get when creating it using get_ordinal_encoder()
```
Comment (bejaeger) [inline src/tabpfn/utils.py]: I would prefer an assert rather than a comment?

File: src/tabpfn/utils.py [hunk 2]
```diff
@@ -627,6 +627,22 @@ def process_text_na_dataframe(
...
+        cat_cols = next(
+            columns
+            for _, transformer, columns in ord_encoder.transformers_
+            if isinstance(transformer, OrdinalEncoder)
+        )
+        num_cols = [c for c in X.columns if c not in cat_cols]
+        col_to_pos = {c: i for i, c in enumerate(cat_cols + num_cols)}
+        X_encoded = X_encoded[:, [col_to_pos[c] for c in X.columns]]
+
```
Comment (bejaeger) [inline src/tabpfn/utils.py]: Hm...this looks a bit hacky and is not easy to generalize to other places.
I think it would be great to have something like a `OrderPreservingColumnTransformer` class that handles that for us and which we instantiate in `get_ordinal_encoder()`. Then we can also switch to this class in other places where we might not treat this properly. What do you think?

File: tests/test_utils.py [hunk 3]
```diff
@@ -192,3 +204,131 @@ def test__infer_devices__device_selected_twice__raises() -> None:
...
+    assert X_out.shape[1] == X.shape[1]
+
+    # Ensure no extra features
+    assert X_out.shape[1] == len(X.columns)
+
+    # Column order: verify string/object columns
+    string_cols = X.select_dtypes(include=["object", "string"]).columns
+    for col in string_cols:
+        col_idx = X.columns.get_loc(col)
+        mask = X[col] == NA_PLACEHOLDER
```
Comment (bejaeger) [inline tests/test_utils.py]: I'm a fan of explicit testing because it is also a great way of documenting. 
Could you add another array of `expected_results` similar to `test_cases` with exactly the `X_out` we expect?
You can maybe remove one or two examples to test because some seem to test similar scenarios.

File: tests/test_utils.py [hunk 4]
```diff
@@ -192,3 +204,131 @@ def test__infer_devices__device_selected_twice__raises() -> None:
...
+    # Infer categorical features
+    cls.inferred_categorical_indices_ = infer_categorical_features(
+        X=X,
+        provided=getattr(cls, "categorical_features_indices", None),
+        min_samples_for_inference=ModelInterfaceConfig.MIN_NUMBER_SAMPLES_FOR_CATEGORICAL_INFERENCE,
+        max_unique_for_category=ModelInterfaceConfig.MAX_UNIQUE_FOR_CATEGORICAL_FEATURES,
+        min_unique_for_numerical=ModelInterfaceConfig.MIN_UNIQUE_FOR_NUMERICAL_FEATURES,
+    )
+
+    # Fix dtypes
```
Comment (bejaeger) [inline tests/test_utils.py]: Could you remove all the comments that are a bit redundant because they just repeat what's written in actual code?

File: src/tabpfn/preprocessors/preprocessing_helpers.py [hunk 5]
```diff
@@ -166,6 +180,79 @@ def transform(self, X: np.ndarray) -> TransformResult:
...
+
+                # map original columns to indices in the transformed array
+                transformed_columns = col_subset + [
+                    c for c in original_columns if c not in col_subset
+                ]
+                indices = [transformed_columns.index(c) for c in original_columns]
+                X = X.iloc[:, indices] if isinstance(X, pd.DataFrame) else X[:, indices]
+                # Once we restored the original columns, for one coder we can break out
+                # of the loop
+                break
```
Comment (bejaeger) [inline src/tabpfn/preprocessors/preprocessing_helpers.py]: What if there are 3 transformers, each of them processing only a subset of columns?
I don't think we have that case anywhere but I think we should change this class to either support the very general case or make sure we don't use it when there are more than 2 transformers (raise ValueError during init).

**Review by bejaeger — COMMENTED** [review]
Thanks @MagnusBuehler , looks great!
I added another comment which would be good to address.

Then another thing, could you add tests specifically for the new Transformer? Would make it easy to ensure it works for different scenarios.

File: tests/test_utils.py [hunk 6]
```diff
@@ -192,3 +204,180 @@ def test__infer_devices__device_selected_twice__raises() -> None:
...
+        X,
+        placeholder=NA_PLACEHOLDER,
+        ord_encoder=ord_encoder,
+        fit_encoder=True,
+    )
+
+    assert X_out.shape[0] == ground_truth.shape[0]
+    assert X_out.shape[1] == ground_truth.shape[1]
+
+    for col_name in X.columns:
```
Comment (bejaeger) [inline tests/test_utils.py:348]: Why don't compare the full `X_out` with the ground truth? am I missing something?

File: src/tabpfn/utils.py [hunk 7]
```diff
@@ -408,7 +415,7 @@ def get_ordinal_encoder(
     # using a regex on the type of the column, and using `object`, `"object"` and
     # `np.object` will not pick up strings.
     to_convert = ["category", "string"]
-    return ColumnTransformer(
+    return OrderPreservingColumnTransformer(
```
Comment (noahho) [inline src/tabpfn/utils.py]: Can we add a larger comment on why we need the order preserving transformer here? Also a note that this might be not necessary if we switch to fully detecting the categoricals based on data type

File: src/tabpfn/preprocessors/preprocessing_helpers.py (hunk 5 above)
Comment (MagnusBuehler) [reply src/tabpfn/preprocessors/preprocessing_helpers.py]: @bejaeger You are right, this should be handled. One issue is that we can only maintain the original column order if all transformers inside the `ColumnTransformer` are subtypes of `OneToOneFeatureMixin` (i.e., have a one-to-one correspondence between input and output features) and all inner transformers process disjoint subsets of features. If two inner transformers (for example, `StandardScaler` and `PowerTransformer`) are applied to the same feature, the `ColumnTransformer` will not pipe them sequentially. Instead, it will create one output feature with `StandardScaler` applied and another output feature with `PowerTransformer` applied. This increases the total number of output features by one, which would require updating `inferred_categorical_variables` accordingly. The same applies to other non-`OneToOneFeatureMixin` transformers, such as `OneHotEncoder`. 
I think as a quick fix raising a `ValueError` during init would work for now.

File: tests/test_utils.py (hunk 6 above)
Comment (MagnusBuehler) [reply tests/test_utils.py:348]: I do not think we can compare the full `X_out` with the ground truth if we use the `get_ordinal_encoder` function, which does not pass a specific `categories` parameter to control for the exact encoding.
 For example,  for a column with entries `["cold", "warm", "cold", "hot"]`, the encoder could map these to `[0, 1, 0, 2]`, `[1, 0, 1, 2]`, or other permutations depending on the internal ordering of categories. 
There is no guarantee which numeric label corresponds to which category. Meanwhile, the ground truth represents one specific manual mapping, such as `[2, 1, 2, 0]`. As a result, even if the encoding is technically correct, it may not match the ground truth exactly.

Comment (bejaeger) [reply tests/test_utils.py:348]: Hm...okay, but isn't the `OrdinalEncoder` deterministic in mapping objects to ints? Meaning you can setup a test case for it?

**Review by bejaeger — APPROVED** [review]
Cool! Thanks for working on this and implementing all the changes.
LGTM let's get this in :)

**Comment by LeoGrin** [comment]
@MagnusBuehler is this ready to merge?

File: src/tabpfn/preprocessors/preprocessing_helpers.py [hunk 8]
```diff
@@ -166,6 +185,121 @@ def transform(self, X: np.ndarray) -> TransformResult:
...
+        **kwargs : additional keyword arguments
+            Passed to sklearn.compose.ColumnTransformer.
+        """
+        super().__init__(transformers=transformers, **kwargs)
+
+        # Check if there is a single transformer, of subtype OneToOneFeatureMixin
+        assert all(
+            isinstance(t, OneToOneFeatureMixin)
+            for name, t, _ in transformers
+            if name != "remaineder"
```
[the remark this answers is not in the document]
Comment (LeoGrin) [reply src/tabpfn/preprocessors/preprocessing_helpers.py]: +1

File: src/tabpfn/utils.py (hunk 7 above)
Comment (MagnusBuehler) [reply src/tabpfn/utils.py]: done

File: src/tabpfn/preprocessors/preprocessing_helpers.py [hunk 9]
```diff
@@ -7,10 +7,29 @@
...
+
+import pandas as pd
+from sklearn.base import (
+    BaseEstimator,
+    check_is_fitted,
+)
+from sklearn.compose import ColumnTransformer, make_column_selector
+from sklearn.preprocessing import FunctionTransformer, OrdinalEncoder
+
+if TYPE_CHECKING:
```
Comment (noahho) [inline src/tabpfn/preprocessors/preprocessing_helpers.py]: Can this be merged with above TYPE_CHECKING?

Comment (noahho) [reply src/tabpfn/preprocessors/preprocessing_helpers.py]: the import section just be cleaned up a bit (reorder)

**Review by noahho — APPROVED** [review]
LGTM, @MagnusBuehler please merge if ready

File: src/tabpfn/preprocessors/preprocessing_helpers.py (hunk 9 above)
Comment (MagnusBuehler) [reply src/tabpfn/preprocessors/preprocessing_helpers.py]: good catch! I organized the imports.
